mypy/plugins/dataclasses.py

```diff
@@ -386,3 +395,100 @@ def _collect_field_args(expr: Expression) -> Tuple[bool, Dict[str, Expression]]:
+                    # `dict` as dict_factory, so we can't special-case that.
+                    return ctx.default_return_type
+
+    ctx.api.fail("'dataclasses.asdict' should be called on dataclass instances",
+                 ctx.context)
+    return ctx.default_return_type
+
+
+def _transform_type_args(*, typ: Instance, transform: Callable[[Instance], Type]) -> List[Type]:
+    """For each type arg used in the Instance, call transform on it if the arg is an Instance."""
+    new_args = []
+    for arg in typ.args:
+        proper_arg = get_proper_type(arg)
+        if isinstance(proper_arg, Instance):
+            new_args.append(transform(proper_arg))
+        else:
+            new_args.append(arg)
+    return new_args
+
+
+def _asdictify(api: CheckerPluginInterface, context: Context, typ: Type) -> Type:
+    """Convert dataclasses into TypedDicts, recursively looking into built-in containers.
+
+    It will look for dataclasses inside of tuples, lists, and dicts and convert them to TypedDicts.
+    """
+
+    def _asdictify_inner(typ: Type, seen_dataclasses: FrozenSet[str]) -> Type:
+        typ = get_proper_type(typ)
+        if isinstance(typ, UnionType):
+            return UnionType([_asdictify_inner(item, seen_dataclasses) for item in typ.items])
+        if isinstance(typ, Instance):
+            info = typ.type
+            if is_type_dataclass(info):
+                if info.fullname in seen_dataclasses:
+                    api.fail(
+                        "Recursive types are not supported in call to asdict, so falling back to "
+                        "Dict[str, Any]",
+                        context)
+                    # Note: Would be nicer to fallback to default_return_type, but that is Any
+                    # (due to overloads?)
+                    return api.named_generic_type('builtins.dict',
+                                                  [api.named_generic_type('builtins.str', []),
+                                                   AnyType(TypeOfAny.implementation_artifact)])
+                seen_dataclasses |= {info.fullname}
+                attrs = info.metadata['dataclass']['attributes']
+                fields = OrderedDict()  # type: OrderedDict[str, Type]
+                for data in attrs:
+                    attr = DataclassAttribute.deserialize(info, data, api)
+                    sym_node = info.names[attr.name]
+                    attr_type = sym_node.type
+                    assert attr_type is not None
+                    fields[attr.name] = _asdictify_inner(attr_type, seen_dataclasses)
+                return make_anonymous_typeddict(api, fields=fields,
+                                                required_keys=set())
+            elif info.has_base('builtins.list'):
+                supertype_instance = map_instance_to_supertype(typ, api.named_generic_type(
+                    'builtins.list', []).type)
```

Comment:
Creating instance with wrong type argument count (event for very short time) is a bad idea. Use dummy `Any` here.

Reply:
Makes sense. Out of curiosity, is there any reason why this is not checked in an assertion? 
Looks like the code for `named_generic_type` has a TODO comment:
```
        # TODO: assert len(args) == len(info.defn.type_vars)
```
Is it because of performance reasons?